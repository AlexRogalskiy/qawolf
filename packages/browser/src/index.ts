export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
