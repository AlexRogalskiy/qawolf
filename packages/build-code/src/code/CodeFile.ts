// virtualizable/mockable
export class CodeFile {
  // load or create initial
  // ----
  // save
  // ----
  // revert

  async init() {
    // const codeExists = await pathExists(this._options.codePath);
    // if (codeExists) {
    //   this.load()
    // }
  }

  async load() {
    // TODO....
    // const code = await readFile(this._options.codePath, "utf8");
  }

  async createInitialCode() {
    // const initialCode = buildInitialCode({
    //   ...this._options,
    //   createCodeSymbol: CREATE_CODE_SYMBOL
    // });
    // await outputFile(this._options.path, initialCode, "utf8");
  }

  discard() {
    // if (this._preexistingCode) {
    //   console.log(yellow("reverted:"), this._options.codePath);
    //   await outputFile(this._options.codePath, this._preexistingCode, "utf8");
    // } else {
    //   await remove(this._options.codePath);
    // }
  }
}
