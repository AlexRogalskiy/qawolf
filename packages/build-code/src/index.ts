export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
