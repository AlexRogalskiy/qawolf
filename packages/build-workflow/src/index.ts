// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
