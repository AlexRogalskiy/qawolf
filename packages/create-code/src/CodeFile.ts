import {
  buildStepsCode,
  BuildStepsOptions,
  buildInitialCode,
  InitialCodeOptions
} from "@qawolf/build-code";
import { Step } from "@qawolf/types";
import { outputFile, pathExists, readFile, remove } from "fs-extra";
import { relative } from "path";
import { PATCH_HANDLE, patchCode } from "./patchCode";
import { removeLinesIncluding } from "./format";

type ConstructorOptions = Omit<InitialCodeOptions, "patchHandle"> & {
  path: string;
};

type PatchOptions = {
  removeHandle?: boolean;
  steps: Step[];
};

export class CodeFile {
  private _commitedStepIndex: number = 0;
  private _isTest: boolean;
  private _lock: boolean;
  private _name: string;
  private _path: string;
  protected _preexisting: string | undefined;

  protected constructor({ isTest, name, path }: ConstructorOptions) {
    this._isTest = !!isTest;
    this._name = name;
    this._path = path;
  }

  public static async loadOrCreate(options: ConstructorOptions) {
    const file = new CodeFile(options);

    file._preexisting = await loadFileIfExists(options.path);

    if (!file._preexisting) {
      await createInitialCode(options);
    }

    return file;
  }

  public async discard() {
    if (this._preexisting) {
      await outputFile(this._path, this._preexisting, "utf8");
    } else {
      await remove(this._path);
    }
  }

  public name() {
    return this._name;
  }

  async patch(options: PatchOptions) {
    if (this._lock) {
      // do not conflict with a patch in progress
      return;
    }

    // patch non-committed (new) steps
    const stepsToPatch = options.steps.slice(this._commitedStepIndex);
    if (!stepsToPatch.length) {
      // no new steps to patch
      return;
    }

    const code = await loadFileIfExists(this._path);
    if (!code) {
      throw new Error("No code to patch");
    }

    this._lock = true;

    let patch = buildPatch({
      isTest: this._isTest,
      steps: stepsToPatch
    });

    let patchedCode = patchCode({ code, patch });

    if (options.removeHandle) {
      patchedCode = removeLinesIncluding(patchedCode, PATCH_HANDLE);
    }

    await outputFile(this._path, patchedCode, "utf8");

    this._commitedStepIndex += stepsToPatch.length;
    this._lock = false;
  }

  public relativePath() {
    return relative(process.cwd(), this._path);
  }
}

export const buildPatch = (options: BuildStepsOptions) => {
  let patch = buildStepsCode(options);

  // include the patch symbol so we can replace it later
  patch += PATCH_HANDLE;

  return patch;
};

export const createInitialCode = async (options: ConstructorOptions) => {
  const initialCode = buildInitialCode({
    ...options,
    patchHandle: PATCH_HANDLE
  });

  await outputFile(options.path, initialCode, "utf8");
};

export const loadFileIfExists = async (
  path: string
): Promise<string | undefined> => {
  const codeExists = await pathExists(path);
  if (!codeExists) return undefined;

  const file = await readFile(path, "utf8");
  return file;
};
