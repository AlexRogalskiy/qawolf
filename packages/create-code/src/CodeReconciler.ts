import { VirtualCode } from "@qawolf/build-code";
import { patchCode, PATCH_HANDLE } from "./patchCode";

type ReconcileOptions = {
  actualCode: string;
  virtualCode: VirtualCode;
};

export class CodeReconciler {
  // the virtual representation of the current code
  private _virtualCode: VirtualCode = new VirtualCode([]);

  private _insertNewExpressions({ actualCode, virtualCode }: ReconcileOptions) {
    const newExpressions = this._virtualCode.newExpressions(virtualCode);
    if (newExpressions.length < 1) return actualCode;

    const patch =
      newExpressions.map(expression => expression.code()).join("") +
      PATCH_HANDLE;

    return patchCode({ code: actualCode, patch });
  }

  private _updateLastExpression({ actualCode, virtualCode }: ReconcileOptions) {
    /**
     * Update the last expression if it changed.
     */
    const codeToUpdate = this._virtualCode.codeToUpdate(virtualCode);
    if (!codeToUpdate) return actualCode;

    // find the last occurrence of the original expression
    const indexToReplace = actualCode.lastIndexOf(codeToUpdate.original);

    // we cannot find the original expression so return the unmodified code
    if (indexToReplace < 0) return actualCode;

    const updatedCode =
      actualCode.slice(0, indexToReplace) +
      actualCode
        .slice(indexToReplace)
        .replace(codeToUpdate.original, codeToUpdate.updated);

    return updatedCode;
  }

  public hasChanges(virtualCode: VirtualCode) {
    return (
      this._virtualCode.codeToUpdate(virtualCode) ||
      this._virtualCode.newExpressions(virtualCode).length > 0
    );
  }

  public reconcile({ actualCode, virtualCode }: ReconcileOptions): string {
    let updatedCode = this._updateLastExpression({ actualCode, virtualCode });

    updatedCode = this._insertNewExpressions({
      actualCode: updatedCode,
      virtualCode
    });

    return updatedCode;
  }

  public update(virtualCode: VirtualCode) {
    this._virtualCode = virtualCode;
  }
}
