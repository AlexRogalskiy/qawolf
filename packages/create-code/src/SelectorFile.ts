import { stepToSelector } from "@qawolf/build-code";
import { Selector, Step } from "@qawolf/types";
import { pathExists, readJson, outputJson, remove } from "fs-extra";
import { concat } from "lodash";

type ConstructorOptions = {
  path: string;
};

type PatchOptions = {
  steps: Step[];
};

export class SelectorFile {
  private _path: string;

  // public for tests
  public _preexistingSelectors: Selector[];
  private _newSelectors: Selector[] = [];

  private _lock: boolean;

  constructor(options: ConstructorOptions) {
    this._path = options.path;
  }

  protected async _write() {
    await outputJson(this._path, this.selectors(), { spaces: " " });
  }

  public static async loadOrCreate(options: ConstructorOptions) {
    const file = new SelectorFile(options);
    file._preexistingSelectors = await loadSelectors(options.path);

    if (!file._preexistingSelectors) {
      await file._write();
    }

    return file;
  }

  public async discard() {
    if (this._preexistingSelectors) {
      await outputJson(this._path, this._preexistingSelectors, {
        spaces: " "
      });
    } else {
      await remove(this._path);
    }
  }

  public hasPreexisting() {
    return !!this._preexistingSelectors;
  }

  public async patch(options: PatchOptions) {
    if (this._lock || options.steps.length <= this._newSelectors.length) {
      return;
    }

    this._lock = true;

    // we do not support editing of the selectors file
    // so we can replace the new selectors
    this._newSelectors = options.steps.map(step => ({
      ...stepToSelector(step),
      // inline index so it is easy to correlate with the test
      index: step.index + (this._preexistingSelectors || []).length
    }));

    await this._write();

    this._lock = false;
  }

  public selectors(): Selector[] {
    return concat(this._preexistingSelectors || [], this._newSelectors);
  }
}

export const loadSelectors = async (path: string) => {
  const codeExists = await pathExists(path);
  if (!codeExists) return;

  const file = await readJson(path);
  return file;
};
