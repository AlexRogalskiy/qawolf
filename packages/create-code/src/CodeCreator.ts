import { StepBuilder } from "@qawolf/build-code";
import { ElementEvent } from "@qawolf/types";
import { bold } from "kleur";
import { throttle } from "lodash";
import { CodeFile } from "./CodeFile";
import { PATCH_HANDLE } from "./patchCode";
import { SelectorFile } from "./SelectorFile";

type ConstructorOptions = {
  codeFile: CodeFile;
  isTest?: boolean;
  selectorFile: SelectorFile;
};

type StartOptions = {
  codePath: string;
  device?: string;
  isTest?: boolean;
  name: string;
  selectorPath: string;
  url: string;
};

export class CodeCreator {
  private _codeFile: CodeFile;
  private _isTest: boolean;
  private _pollingIntervalId?: NodeJS.Timeout;
  private _selectorFile: SelectorFile;
  private _stepBuilder: StepBuilder;

  protected constructor({
    codeFile,
    isTest,
    selectorFile
  }: ConstructorOptions) {
    this._codeFile = codeFile;
    this._isTest = !!isTest;
    this._selectorFile = selectorFile;

    this._stepBuilder = new StepBuilder({
      startIndex: this._selectorFile.selectors().length
    });
  }

  public static async start(options: StartOptions) {
    const codeFile = await CodeFile.loadOrCreate({
      ...options,
      path: options.codePath
    });

    const selectorFile = await SelectorFile.loadOrCreate({
      path: options.selectorPath
    });
    return new CodeCreator({ codeFile, isTest: options.isTest, selectorFile });
  }

  private _logNoHandle = throttle(
    () => {
      console.warn(
        "\n",
        bold().red("Cannot update code without this line:"),
        PATCH_HANDLE
      );
    },
    10000,
    { leading: true }
  );

  private _logSaveSuccess() {
    const command = this._isTest
      ? `npx qawolf test ${this._codeFile.name()}`
      : `node ${this._codeFile.relativePath()}`;

    console.log(
      bold().blue(`✨  Created your ${this._isTest ? "test" : "script"}`)
    );
    console.log(bold().blue("🐺  Run it with:"), command);
  }

  private async _patchFiles(removeHandle: boolean = false) {
    const steps = this._stepBuilder.steps();
    await Promise.all([
      this._codeFile.patch({ removeHandle, steps }),
      this._selectorFile.patch({ steps })
    ]);
  }

  public async discard() {
    await Promise.all([this._codeFile.discard(), this._selectorFile.discard()]);
  }

  public pushEvent(event: ElementEvent) {
    this._stepBuilder.pushEvent(event);
  }

  public async save() {
    this._stepBuilder.finalize();
    await this._patchFiles(true);
    this._logSaveSuccess();
  }

  public startPolling() {
    this._pollingIntervalId = setInterval(async () => {
      try {
        await this._patchFiles();
      } catch (e) {
        if (e.message.includes("Cannot patch without handle")) {
          this._logNoHandle();
        } else {
          throw e;
        }
      }
    }, 100);
  }

  public stopPolling() {
    if (!this._pollingIntervalId) return;

    clearInterval(this._pollingIntervalId);
    this._pollingIntervalId = undefined;
  }
}
