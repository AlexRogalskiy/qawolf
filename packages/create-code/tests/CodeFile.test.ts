import { outputFile, pathExists, readFile, remove } from "fs-extra";
import { CodeFile, CodeFileOptions } from "../src/CodeFile";

// require manually since fs is mocked
const scrollLogin = require("@qawolf/fixtures/workflows/scroll_login.json");

jest.mock("fs-extra");

const mockedPathExists = pathExists as jest.Mock;
const mockedReadFile = readFile as jest.Mock;
const mockedRemove = remove as jest.Mock;
const mockedOutputFile = outputFile as jest.Mock;

const options: CodeFileOptions = {
  path: "/path/tests/mytest.test.js",
  isTest: true,
  name: "mytest",
  url: "http://localhost:3000"
};

describe("CodeFile", () => {
  describe("preexisting code", () => {
    let codeFile: CodeFile;
    let preexisting = "preexistingCode()";

    beforeAll(async () => {
      mockedPathExists.mockResolvedValueOnce(true);
      mockedReadFile.mockResolvedValueOnce(preexisting);
      mockedRemove.mockReset();
      mockedOutputFile.mockReset();
      codeFile = await CodeFile.loadOrCreate(options);
    });

    it("loadOrCreate: loads preexisting code", () => {
      expect(codeFile._preexisting).toEqual(preexisting);
    });

    it("discard: reverts to preexisting code", async () => {
      await codeFile.discard();
      expect(mockedRemove.mock.calls.length).toEqual(0);
      expect(mockedOutputFile.mock.calls.length).toEqual(1);
      expect(mockedOutputFile.mock.calls[0]).toEqual([
        options.path,
        preexisting,
        "utf8"
      ]);
    });
  });

  describe("no preexisting code", () => {
    let codeFile: CodeFile;

    beforeAll(async () => {
      mockedOutputFile.mockReset();
      codeFile = await CodeFile.loadOrCreate(options);
    });

    it("loadOrCreate: creates code", () => {
      expect(mockedOutputFile.mock.calls.length).toEqual(1);
      expect(mockedOutputFile.mock.calls[0]).toMatchSnapshot();
    });

    it("discard: deletes the new code", async () => {
      mockedRemove.mockReset();
      mockedOutputFile.mockReset();
      await codeFile.discard();
      expect(mockedRemove.mock.calls.length).toEqual(1);
      expect(mockedOutputFile.mock.calls.length).toEqual(0);
    });
  });

  describe("patch", () => {
    it("saves a patch for non-committed steps", async () => {
      mockedOutputFile.mockReset();

      const codeFile = await CodeFile.loadOrCreate(options);

      mockedPathExists.mockResolvedValue(true);

      const initialFile = mockedOutputFile.mock.calls[0][1];
      mockedReadFile.mockResolvedValueOnce(initialFile);

      await codeFile.patch({
        steps: scrollLogin.steps.slice(0, 2)
      });

      const fileRevisionOne = mockedOutputFile.mock.calls[1][1];
      expect(fileRevisionOne).toMatchSnapshot();
      mockedReadFile.mockResolvedValueOnce(fileRevisionOne);

      await codeFile.patch({
        removeHandle: true,
        steps: scrollLogin.steps
      });

      const fileRevisionTwo = mockedOutputFile.mock.calls[2][1];
      expect(fileRevisionTwo).toMatchSnapshot();
    });
  });
});
