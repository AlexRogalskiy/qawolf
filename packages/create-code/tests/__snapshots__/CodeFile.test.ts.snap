// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeFile no preexisting code loadOrCreate: creates code 1`] = `
Array [
  "/path/tests/mytest.test.js",
  "const { launch } = require(\\"qawolf\\");
const selectors = require(\\"../selectors/mytest\\");

describe('mytest', () => {
  let browser;

  beforeAll(async () => {
    browser = await launch({ url: \\"http://localhost:3000\\" });
  });

  afterAll(() => browser.close());
  // 🐺 CREATE CODE HERE
});",
  "utf8",
]
`;

exports[`CodeFile patch saves a patch for non-committed steps 1`] = `
"const { launch } = require(\\"qawolf\\");
const selectors = require(\\"../selectors/mytest\\");

describe('mytest', () => {
  let browser;

  beforeAll(async () => {
    browser = await launch({ url: \\"http://localhost:3000\\" });
  });

  afterAll(() => browser.close());
  
  it('can scroll', async () => {
    await browser.scroll(selectors[0], { x: 0, y: 378 });
  });
  
  it('can click \\"Form Authentication\\" link', async () => {
    await browser.click(selectors[1]);
  });
  // 🐺 CREATE CODE HERE
});"
`;

exports[`CodeFile patch saves a patch for non-committed steps 2`] = `
"const { launch } = require(\\"qawolf\\");
const selectors = require(\\"../selectors/mytest\\");

describe('mytest', () => {
  let browser;

  beforeAll(async () => {
    browser = await launch({ url: \\"http://localhost:3000\\" });
  });

  afterAll(() => browser.close());
  
  it('can scroll', async () => {
    await browser.scroll(selectors[0], { x: 0, y: 378 });
  });
  
  it('can click \\"Form Authentication\\" link', async () => {
    await browser.click(selectors[1]);
  });
  
  it('can type into \\"Username\\" input', async () => {
    await browser.type(selectors[2], \\"tomsmith\\");
  });
  
  it('can Tab', async () => {
    await browser.type(selectors[3], \\"↓Tab↑Tab\\");
  });
  
  it('can type into \\"Password\\" input', async () => {
    await browser.type(selectors[4], \\"SuperSecretPassword!\\");
  });
  
  it('can Enter', async () => {
    await browser.type(selectors[5], \\"↓Enter↑Enter\\");
  });
  
  it('can click \\"Logout\\" link', async () => {
    await browser.click(selectors[6]);
  });
});"
`;
